compiler/resolution/wrappers.cpp

```diff
@@ -720,11 +721,89 @@ void coerceActuals(FnSymbol* fn, CallInfo* info) {
 //// promotion wrapper code
 ////
 
+//
+// In order for fast followers to trigger, the invoking loop requires a static
+// and dynamic check. They determine at compile time if the iterands implement
+// a fast follower and at runtime check if all the iterands are can fast follow
+// the leader. Here we build up the checks for an iterator record. We basically
+// convert the iterator record into a tuple and call the "zip" check. Note that
+// we only stuff the components of the iterator record that actually require
+// promotion into the tuple.
+//
+// Since the iterator record isn't fully-built yet, we use a primitive as a
+// placeholder. I gets replaced when the iterator record is populated during
+// iterator lowering.
+//
+static void
+buildPromotionFastFollowerCheck(bool isStatic,
+                                bool zippered,
+                                bool addLead,
+                                CallInfo* info,
+                                FnSymbol* wrapper,
+                                std::set<ArgSymbol*>& requiresPromotion) {
+
+  // a not so elegant way to determine which check to build, and which check
+  // we're forwarding to
+  const char* fnName = isStatic ? "chpl__staticFastFollowCheck" : "chpl__dynamicFastFollowCheck";
+  const char* fowardFnName = astr(fnName, "Zip") ;
+
+  FnSymbol* fastFollowCheckFn = new FnSymbol(fnName);
+  if (isStatic) {
+    fastFollowCheckFn->retTag = RET_PARAM;
+  } else {
+    fastFollowCheckFn->retTag = RET_VALUE;
+  }
+
+  ArgSymbol* x = new ArgSymbol(INTENT_BLANK, "x", dtIteratorRecord);
```

Comment:
How about a meaningful name instead of "x" ?
For the generated code, that is. If also for the compiler code, that will be super.